proposals/p1964.md

```diff
@@ -0,0 +1,356 @@
+character literals will use their own types distinct from these:
+
+-   We will support value preserving implicit conversions from character
+    literals to code point or code unit types. In particular, a character
+    literal converts to a `Char8` UTF-8 code unit if it is less than or equal to
+    0x7F, and `Char16` UTF-16 code unit if it is less than or equal to 0xFFFF.
+-   Conversions from string or character literals to a non-value-preserving
+    encoding must be explicit.
+-   Conversions from string literals to Unicode strings are implicit, even
+    though the numeric values of the encoding may change.
+
+We can see whether the particular literal is represented in the variable's type
+by only looking at the types.
+
+```
+let allowed: Char8 = 'a';
+```
+
+The above is allowed because the type of `'a'` is the character literal
+consisting of the single Unicode code point 97, which can be converted to
+`Char8` since 97 is less than or equal to 0x7F.
+
+```
+let error1: Char8 = '😃';
+let error2: Char8 = 'AB';
+```
+
+However these should produce errors. The type of `'😃'` is the character literal
+consisting of the single Unicode code point `0x1F603`, which is greater than
+0x7F. The type of `'AB'` is a character literal that is a sequence of two
+Unicode code points, which is not valid.
+
+Literals `'\n'` and `'\u{A}'` represent the same character and so have the same
+type. However, explicitly converting this character literal to another character
+set might result in a character with a different value, but that still
+represents the newline character.
+
+```
+// 0x15 is the new line character in EBCDIC.
+Assert('\n' as EBCDICChar == EBCDICChar.Make(0x15));
+Assert('\u{A}' as EBCDICChar != EBCDICChar.Make(0x0A));
+```
+
+### Operations
+
+Character literals representing a single code point support the following
+operators:
+
+-   Comparison: `<`, `>`, `<=`, `>=` `==`
+-   Plus: `+`. However the behavior of this operator is different from its use
+    with strings. Rather than a concatenation, this will add the value of the
+    two characters:
+    -   If the `+` is used between a character literal representing a single
+        Unicode code point and an integer literal, this should produce a
+        character literal if the result fits in a Unicode code point.
+    -   If the `+` is used between two character literals this will produce an
+        error.
+-   Subtract: `-` Similar to the plus operator, this will subtract the value of
+    the two characters.
+
+    -   If the `-` is used between a character literal representing a single
+        code point or code unit and an integer literal, this should produce a
+        character literal as long as the result of the difference is in range.
+    -   If the `-` is used between two character literals, which are both a
+        single code point, this should produce an integer literal representing
+        the difference between the code points.
```

Comment:
I don't think this is similar enough to plus for that to be helpful, so I would just describe this from scratch.

For `-`, I feel like `'z' - 4` makes sense, but `100 - 'a'` doesn't really, especially when producing a character literal type. For me this is because a negative character literal doesn't make much sense (unlike an integer literal). I'd be happy skipping subtraction entirely or making it non-commutative, thoughts?

Reply:
Just to check, I think that agrees with what I'm saying? Want to make sure I'm not missing an important distinction.

Reply:
I think we agree, since I'm fine with rejecting `100 - 'a'` unless you explicitly cast the `100` to a character type.